sys/net/gnrc/routing/aodvv2/aodv.c

```diff
@@ -0,0 +1,543 @@
+
+    msg_try_send(&msg2, sender_thread);
+    mutex_unlock(&rerr_mutex);
+}
+
+/*
+ * init the multicast address all RREQ and RERRS are sent to
+ * and the local address (source address) of this node
+ */
+static void _init_addresses(void)
+{
+    eui64_t iid;
+
+    /* init multicast address: set to to a link-local all nodes multicast address */
+    ipv6_addr_set_all_nodes_multicast(&_v6_addr_mcast, IPV6_ADDR_MCAST_SCP_LINK_LOCAL);
+    DEBUG("my multicast address is: %s\n",
+        ipv6_addr_to_str(addr_str, &_v6_addr_mcast, IPV6_ADDR_MAX_STR_LEN));
+
+    /* get id (of type eui46_t) of the interface we're sending on */
+    int err_code = gnrc_netapi_get(aodvv2_if_id, NETOPT_IPV6_IID, 0, &iid,
+                       sizeof(eui64_t));
+    if (err_code < 0) {
+        DEBUG("Error: failed to get iid, error code: %d\n", err_code);
+        return;
+    }
+
+    /* Set addr according to our interface id */
+    ipv6_addr_set_aiid(&_v6_addr_local, &iid.uint8[0]);
+    /* Set (global!) prefix */
+    ipv6_addr_init_prefix(&_v6_addr_local, &aodvv2_prefix, aodvv2_prefix_len);
+    /* Add our homemade address to our interface (aodvv2_if_id) */
+    gnrc_ipv6_netif_add_addr(aodvv2_if_id, &_v6_addr_local, aodvv2_prefix_len, 0);
+
+    DEBUG("my src address is:       %s\n",
+        ipv6_addr_to_str(addr_str, &_v6_addr_local, IPV6_ADDR_MAX_STR_LEN));
+
+    /* store src & multicast address as netaddr as well for easy interaction
+     * with oonf based stuff */
+    ipv6_addr_t_to_netaddr(&_v6_addr_local, &na_local);
+    ipv6_addr_t_to_netaddr(&_v6_addr_mcast, &na_mcast);
+
+    /* TODO: do I need this?*/
+    ipv6_addr_set_loopback(&_v6_addr_loopback);
+}
+
+/* Build RREQs, RREPs and RERRs from the information contained in the thread's
+ * message queue and send them */
+static void *_aodv_sender_thread(void *arg)
+{
+    (void) arg;
+
+    msg_t snd_msgq[RCV_MSG_Q_SIZE];
+    msg_init_queue(snd_msgq, RCV_MSG_Q_SIZE);
+
+    while (true) {
+        msg_t msg;
+        msg_receive(&msg);
+        struct msg_container *mc = (struct msg_container *) msg.content.ptr;
+
+        if (mc->type == RFC5444_MSGTYPE_RREQ) {
+            struct rreq_rrep_data *rreq_data = (struct rreq_rrep_data *) mc->data;
+            aodv_packet_writer_send_rreq(rreq_data->packet_data, rreq_data->next_hop);
+        }
+        else if (mc->type == RFC5444_MSGTYPE_RREP) {
+            struct rreq_rrep_data *rrep_data = (struct rreq_rrep_data *) mc->data;
+            aodv_packet_writer_send_rrep(rrep_data->packet_data, rrep_data->next_hop);
+        }
+        else if (mc->type == RFC5444_MSGTYPE_RERR) {
+            struct rerr_data *rerr_data = (struct rerr_data *) mc->data;
+            aodv_packet_writer_send_rerr(rerr_data->unreachable_nodes, rerr_data->len,
+                             rerr_data->hoplimit, rerr_data->next_hop);
+        }
+        else {
+            AODV_DEBUG("ERROR: Couldn't identify Message\n");
+        }
+        _deep_free_msg_container(mc);
```

Comment:
this would also require tweak to `free` the content, either by passing the `msg.type` as parameter or just passing the `msg *` and dispatch.